### PR TITLE
Add Ory Kratos Integration.

### LINKED ISSUE #193: [ Task ] Add integration of Kratos User Identity management system.
## Include Ory Kratos support 
- [ ] It should be optional, i.e., the user can enable it if they want.
- [ ] There will be one variable inside `.env` named as `KRATOS_REQUIRED`. If user set it to true support for kratos will be enabled.
- [ ] User will have to add all config inside the `.env` file (reference for the same will be there in `.env.example` )
- [ ] In the auth middleware we've got to check for the kratos cookie and insert it into DB if it''s not already there.
- [ ] Add kratos service into the same docker compose file, will use the profile while running the docker compose to start kratos service.

## Acceptance Criteria
- [ ] The user has to provide the URL of the AUTH UI in config; Kratos will use that url for redirection after creating each flow.
- [ ] for the registration and login flow, after the flow completion, Kratos will redirect to the backend with setting up cookies.
- [ ] Backend will receive session_id from the cookie and store it as a user ID in database (for reference, see the diagram below)

### PR DESCRIPTION
**Pull request by Shaktizala**
Closes #193 

## Changes proposed
- Kratos Integration 
- Endpoint for Kratos redirecting and storing user details in the database.
- Kratos service in docker compose and kratos database in the same PostgreSQL server.

## Check List (Check all the applicable boxes) 

- [x] My code follows the code style of this project.
- [x] My change requires changes to the documentation.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] This PR does not contain plagiarized content.
- [x] The title of my pull request is a short description of the requested changes.

### CODE REVIEW

File: pkg/kratos/readme.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    - For steps of customizing identity schema, visit [here](https://www.ory.sh/docs/kratos/manage-identities/customize-identity-schema).
+
+## Implementation
+For all the endpoints that require Kratos authentication, You'll required to add middleware and auth endpoints before it while establishing routes.
+
+#### Example
+route.Get("/\<end-point\>", middlewares.Authenticated, authController.DoKratosAuth)
+
+## How Kratos Integration Works?
+You have to provide Kratos with your UI endpoint URLs, Kratos will redirect to those endpoints by initializing the flow.
```
Comment (munir131) [inline pkg/kratos/readme.md:67]: @Shaktizala What frontend has to sent? Should it work with ajax request? Does it need to follow any xml/json format? Or it is form?

**Review by munir131 — CHANGES_REQUESTED** [review]

File: pkg/kratos/readme.md (hunk 1 above)
Comment (Shaktizala) [reply pkg/kratos/readme.md:67]: Yes, it is in form.

File: config/kratos.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package config
+
+// DBConfig type of db config object
+type KratosConfig struct {
+	IsRequired bool   `envconfig:"KRATOS_REQUIRED"`
+	BaseURL    string `envconfig:"SERVE_PUBLIC_BASE_URL"`
+	UIUrl      string `envconfig:"SELF_SERVICE_DEFAULT_BROWSER_RETURN_URL"`
+	AdminUrl   string `envconfig:"SERVE_ADMIN_BASE_URL"`
+	PublicUrl  string `envconfig:"SERVE_PUBLIC_BASE_URL"`
```
[the remark this answers is not in the document]
Comment (Shaktizala) [reply config/kratos.go:9]: Both are same, it's the kratos base url, but as kratos requires envs to be in specific format it is kept like, "SERVE_PUBLIC_BASE_URL".

File: controllers/api/v1/auth_controller.go [hunk 3]
```diff
@@ -94,3 +98,61 @@ func (ctrl *AuthController) DoAuth(c *fiber.Ctx) error {
...
+
+	user, err := ctrl.userModel.InsertKratosUser(userStruct)
+	if err != nil {
+		return utils.JSONError(c, http.StatusInternalServerError, constants.ErrKratosAuth)
+	}
+
+	userCookie := &fiber.Cookie{
+		Name:    constants.KratosCookie,
+		Value:   kratosID.(string),
+		Expires: time.Now().Add(24 * time.Hour),
```
[the remark this answers is not in the document]
Comment (Shaktizala) [reply controllers/api/v1/auth_controller.go:149]: For that, we need to fix the format of entering it from env.

[the remark this answers is not in the document]
Comment (Shaktizala) [reply controllers/api/v1/auth_controller.go:149]: or should I take it in hours?

File: controllers/api/v1/auth_controller.go [hunk 4]
```diff
@@ -94,3 +98,61 @@ func (ctrl *AuthController) DoAuth(c *fiber.Ctx) error {
...
+//
+//				Consumes:
+//				- application/json
+//
+//				Schemes: http, https
+//			Responses:
+//		      400: GenericResFailBadRequest
+//			  500: GenericResError
+func (ctrl *AuthController) DoKratosAuth(c *fiber.Ctx) error {
+	kratosID := c.Locals(constants.KratosID)
```
[the remark this answers is not in the document]
Comment (munir131) [reply controllers/api/v1/auth_controller.go:115]: @Shaktizala Is it middleware? Middleware shouldn't be in controller

File: pkg/kratos/readme.md (hunk 1 above)
Comment (munir131) [reply pkg/kratos/readme.md:67]: Then please document that or add link of external doc

File: controllers/api/v1/auth_controller.go [hunk 5]
```diff
@@ -94,3 +98,62 @@ func (ctrl *AuthController) DoAuth(c *fiber.Ctx) error {
...
+		Name:    constants.KratosCookie,
+		Value:   kratosID.(string),
+		Expires: time.Now().Add(cookieExpirationTime),
+	}
+
+	c.Cookie(userCookie)
+	c.Redirect(ctrl.config.Kratos.UIUrl)
+
+	c.Locals(constants.KratosUserDetails, user)
+	c.Next()
```
Comment (munir131) [inline controllers/api/v1/auth_controller.go:157]: @Shaktizala What is purpose of it? L154 will already redirect on UI

File: pkg/kratos/readme.md (hunk 1 above)
Comment (Shaktizala) [reply pkg/kratos/readme.md:67]: okay will update it.

File: controllers/api/v1/auth_controller.go (hunk 5 above)
Comment (Shaktizala) [reply controllers/api/v1/auth_controller.go:157]: sorry, forgot to remove it, it was there when I was not doing redirection. will update it.

**Review by munir131 — DISMISSED** [review]

**Review by pratikbgit — CHANGES_REQUESTED** [review]
- [x] Define step pkg/kratos/oidc then run base64 -w 0 google.schema.jsonnet
- [ ] Mention the steps for OAuth 2.0 client IDs and redirect URLs set.
- [ ] Start service in on go, like Kratos up command then migrate up then api run etc.
- [x] Info: If the password field is not required, then please remove it.

**Comment by Shaktizala** [comment]
> - [ ] Define step pkg/kratos/oidc than run `base64 -w 0 google.schema.jsonnet` 
> - [ ] Mention steps of OAuth 2.0 Client IDs and redirect url set.
> - [ ] Start service in on go like Kratos up command then migrate up then api run etc.
> - [ ] info : If password filed is not required then please remove it.
>  
> 
> 

@pratikbgit Password field is required, when kratos is not enabled We have to use authentication using password so at that time we need to store password.


I am not getting the 3rd statement, can you please elaborate.

CC: @munir131

**Comment by pratikbgit** [comment]
> > * [ ]  Define step pkg/kratos/oidc than run `base64 -w 0 google.schema.jsonnet`
> > * [ ]  Mention steps of OAuth 2.0 Client IDs and redirect url set.
> > * [ ]  Start service in on go like Kratos up command then migrate up then api run etc.
> > * [ ]  info : If password filed is not required then please remove it.
> 
> @pratikbgit Password field is required, when kratos is not enabled We have to use authentication using password so at that time we need to store password.
> 
> I am not getting the 3rd statement, can you please elaborate.
> 
> CC: @munir131

While kratos service start before that migrate and api service should start not manually.

**Comment by Shaktizala** [comment]
> > > * [ ]   Define step pkg/kratos/oidc than run `base64 -w 0 google.schema.jsonnet`
> > > * [ ]   Mention steps of OAuth 2.0 Client IDs and redirect url set.
> > > * [ ]   Start service in on go like Kratos up command then migrate up then api run etc.
> > > * [ ]   info : If password filed is not required then please remove it.
> > 
> > 
> > @pratikbgit Password field is required, when kratos is not enabled We have to use authentication using password so at that time we need to store password.
> > I am not getting the 3rd statement, can you please elaborate.
> > CC: @munir131
> 
> While kratos service start before that migrate and api service should start not manually.

@pratikbgit 
I discussed this with @munir131.
- [ ] We will have to make one script file for starting all three services.
- [ ] There is no need to remove the password field, as it is required when we are not using Kratos.
- [ ] I will add a link to the steps of creating a client ID, secrets, and different details of it.

**Review by munir131 — DISMISSED** [review]

**Comment by munir131** [comment]
@Shaktizala Please resolve conflicts

**Review by munir131 — APPROVED** [review]

**Pull request merged**